File: bipartite.py
# Bipartite Graph Checker

"""
Program to check if given undirected graph is bipartite/2-colourable.
Graph may be disconnected, but has no self edges etc (all the standard rules).
Input 'graph' is an adjacency list.
"""

def isBipartite(graph) -> bool:
    
    n = len(graph)
    colouring = [0]*n # uncoloured = 0, red = +1, blue = -1
    
    # Breadth First Search
    seen = set()
    parity = 1
    
    for i in range(n):
        if i in seen:
            continue
        seen.add(i)
        queue = [i]
        while queue:
            parity *= -1
            next_queue = []
            for current in queue:
                # Colour the node.
                colouring[current] = parity
                # Check this doesn't break anything AND add unvisited adjacents into queue.
                for adjacent in graph[current]:
                    if colouring[adjacent] == parity:
                        # An adjacent node has the same colour - contradiction.
                        return False
                    if adjacent not in seen:
                        next_queue.append(adjacent)
                        seen.add(adjacent)
            queue = next_queue
    # If code reaches this point then we have a valid 2-colouring.
    return True

File: bipartite.cpp
// Bipartite Graph Checker
/*
Program to check if given undirected graph is bipartite/2-colourable.
Graph may be disconnected, but has no self edges etc (all the standard rules).
Input 'graph' is an adjacency list.
*/

#include <vector>
#include <unordered_set>

bool isBipartite(std::vector<std::vector<int>>& graph) {

    int n = graph.size();
    int parity = 1;
    std::vector<int> colouring = std::vector<int>(n);
    std::unordered_set<int> seen;
    std::vector<int> queue;
    std::vector<int> next_queue;

    for (int i = 0; i < n; i++) {

        if (seen.find(i) != seen.end()) continue;
        seen.insert(i);
        queue.clear();
        queue.push_back(i);

        while (!queue.empty()) {

            parity *= -1;
            next_queue.clear();

            for (int current : queue) {

                colouring[current] = parity;
                for (int adjacent : graph[current]) {

                    if (colouring[adjacent] == parity) return false;
                    if (seen.find(adjacent) == seen.end()) {

                        next_queue.push_back(adjacent);
                        seen.insert(adjacent);

                    }

                }

            }

            queue = next_queue;

        }

    }

    return true;

}